book/src/frequently_asked_questions.md

```diff
@@ -37,6 +37,19 @@ practice. This can allow account hijacking, privilege escalation, credential dis
 information leaks and more. The entire path between a client and the server must be protected at all
 times.
 
+There are a variety of ways that you can configure TLS between your load balancer and Kanidm.
+Ultimately, any option that maintains the confidentiality and integrity of the communication will
+suffice. Some options include, but are not limited to:
+
+- Generating a self-signed certificate
```

Comment:
The self-signed certificates generated by Kanidm aren't designed for any production use, so until that changes the help text is correct.

Reply:
They're designed (ref: [build_cert](https://github.com/yaleman/kanidm/blob/5c5da3c867bf18ee220bc611bf151ce72e5b0d99/server/core/src/crypto.rs#L435)) for test/dev purposes only with a 5 day expiry and some simple defaults. 

There's a whole bunch more config that goes into a well-formed certificate, including location, hostnames, contact details etc. If someone wanted to make it easy to generate them then it might be entertained, but ... there's also SO many other software platforms that can do this, and duplicating *production quality* features for the sake of it is something we try to avoid.

Reply:
It depends on the threat model of your environment, like most security decisions. If folks are reading _our_ docs to decide on how to generate a the cert between their app and a load balancer, simple `openssl` commands would do fine. 😄

Reply:
I wonder if we should just make a cert-helper for that specific use case?

Reply:
> on second thought, do they though? does certificate pinning care about inconsistent hostnames or expired certificates

that's up to your load balancer configs.

@Firstyear we could do that, sure

Reply:
If only I was already planning to make a CA ....

Reply:
> Aside from [smallstep](https://github.com/smallstep/certificates) (which just seems like running your own CA), [mkcert](https://github.com/FiloSottile/mkcert) (also makes a local CA, but tries to abstract that away for you)

You don't have to run your own CA service with `smallstep` btw, I would say it's much better than `mkcert` (_at least when I last looked at that and later discovered `smallstep`_).

Running the commands on the host is a little nicer if you want the benefits like updating the OS trust store, but here is a quick and easy way packaged into a single `compose.yaml` file.

```yaml
# compose.yaml example, easy local TLS certificate provisioning

# USAGE:
# Basic example, provision a cert for `example.test` + `*.example.test`:
# docker compose run --rm gen-certs
#
# Custom domain:
# docker compose run --rm gen-certs custom.test
#
# Use the `step` CLI to inspect a certificate at `/tmp/certs`:
# docker compose run --rm --entrypoint /usr/local/bin/step gen-certs certificate inspect cert.ecdsa.pem

services:
  gen-certs:
    image: smallstep/step-ca
    # Your certs will be stored out of the container at `certs/`:
    volumes:
      - "./certs/:/tmp/certs/"
    # Typically if the local host path for the volume above hasn't been created,
    # it'll be created with `root` ownership (unless you use the `run --user` option),
    # The image defaults to a non-root user `step`, reset back to `root` to have permission
    # to write the files to the host volume path. If you already created the directory with
    # your own users ownership, you could override this with: `docker compose run --user "$(id -u):$(id -g)" gen-certs`
    user: root
    # Override the image entrypoint to run our own script instead:
    entrypoint: /opt/scripts/generate-certs.sh
    # From this location:
    working_dir: /tmp/certs
    configs:
      - source: generate-certs
        target: /opt/scripts/generate-certs.sh
        # When relying on a shebang to delegate the shell interpreter to run,
        # you will need the file to have the executable bit (`chmod +x`).
        # Here we assign permissions r-x for the active container user:
        mode: 0500

# This `configs` compose feature is compatible with any version of Docker Compose released since 2024.
# This feature allows you to inline files within `compose.yaml`,
# instead of separate files (added via `volumes` bind mount or custom `Dockerfile`).
# NOTE: `compose.yaml` has a feature that interpolates ENV vars; that is prevented via the escaped form: $$.
#
# Here we use it to run some CLI commands when the container is started.
configs:
  generate-certs:
    content: |
      #!/usr/bin/env bash

      # INFO: Provisions a CA cert and leaf cert offline without requiring ACME:
      # - The private keys have no external secret input, possession of them is sufficient to sign with.
      # - Key type is ECDSA with P-256 curve, fairly modern and secure choice.
      # - These are provisioned for convenience of testing, hence the 10 year validity period. Production should minimize this.
      # - For any connecting clients to your server using the leaf cert, they will need the CA cert in
      #   their trust store (either the OS trust store, or if client software supports it such as curl via CLI option).

      # A custom domain could be provided as an arg, otherwise fallback to `example.test`:
      ARG_FQDN="$${1:-example.test}"

      # A self-signed leaf cert does not externalize trust,
      # that's where a CA cert for your connecting clients helps establish trust.

      # The CA that signs the leaf certs with the CA key.
      # LetsEncrypt for example signs your leaf cert when provisioned,
      # and most systems have a valid LetsEncrypt public CA cert to verify chain of trust
      # during the TLS handshake to secure the connection.
      step certificate create "Smallstep Root CA" ca-cert.ecdsa.pem ca-key.ecdsa.pem \
        --no-password --insecure \
        --profile root-ca \
        --not-before "2024-01-01T00:00:00+00:00" \
        --not-after "2034-01-01T00:00:00+00:00" \
        --kty EC --crv P-256

      # The leaf cert, your more familiar TLS cert + key files.
      # Here we have signed with the CA files created above and assigned two SAN entries,
      # Making it valid for `example.test` and any subdomain of it via wildcard.
      step certificate create "Smallstep Leaf" cert.ecdsa.pem key.ecdsa.pem \
        --no-password --insecure \
        --profile leaf \
        --ca ca-cert.ecdsa.pem \
        --ca-key ca-key.ecdsa.pem \
        --not-before "2024-01-01T00:00:00+00:00" \
        --not-after "2034-01-01T00:00:00+00:00" \
        --san "$${ARG_FQDN}" \
        --san "*.$${ARG_FQDN}" \
        --kty EC --crv P-256
```

See the example commands at the start of that config. You just need a Docker Compose release from 2024 installed and to copy/paste that content to a `compose.yaml` file locally, then a simple command will create the files 👍 

It does look quite verbose since I bundled the shell script and heavily documented it, but it's really minimal in practice:

```yaml
# Provision for `custom.test` + `*.custom.test`:
# docker compose run --rm gen-certs custom.test

# `compose.yaml` content:
services:
  gen-certs:
    image: smallstep/step-ca
    volumes:
      - "./certs/:/tmp/certs/"
    user: root
    entrypoint: /opt/scripts/generate-certs.sh
    working_dir: /tmp/certs
    configs:
      - source: generate-certs
        target: /opt/scripts/generate-certs.sh
        mode: 0500

configs:
  generate-certs:
    content: |
      #!/usr/bin/env bash

      ARG_FQDN="$${1:-example.test}"

      step certificate create "Smallstep Root CA" ca-cert.ecdsa.pem ca-key.ecdsa.pem \
        --no-password --insecure \
        --profile root-ca \
        --not-before "2024-01-01T00:00:00+00:00" \
        --not-after "2034-01-01T00:00:00+00:00" \
        --kty EC --crv P-256

      step certificate create "Smallstep Leaf" cert.ecdsa.pem key.ecdsa.pem \
        --no-password --insecure \
        --profile leaf \
        --ca ca-cert.ecdsa.pem \
        --ca-key ca-key.ecdsa.pem \
        --not-before "2024-01-01T00:00:00+00:00" \
        --not-after "2034-01-01T00:00:00+00:00" \
        --san "$${ARG_FQDN}" \
        --san "*.$${ARG_FQDN}" \
        --kty EC --crv P-256
```

For production usage, you could do similar with `certbot` or a reverse-proxy that handles the same thing with LetsEncrypt with less lines and automates updating before cert expires.

If you do need the cert to be managed externally of a reverse-proxy due to no support from the service to delegate TLS termination, you should be able to proxy a passthrough connection. This is Layer 7 with TCP instead of Layer 4 protocols (HTTP), so the proxy is a bit more limited in what it can do, you may also lose the original connection IP this way if the service does not support PROXY protocol headers or some other way of persisting that information (_non-issue for HTTPS traffic due to header support there_)

Reply:
Didn't you change the expiries on the self-generated certs, or was that only for mTLS?